// @geckos.io/phaser-on-nodejs/lib/fakeXMLHttpRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
class FakeXMLHttpRequest {
    constructor() {
        this.status = 200;
    }
    open(_type, url) {
        this.url = path_1.default.resolve(__dirname, url);
    }
    send() {
        // use base64 for images and utf8 for json files
        const encoding = /\.json$/gm.test(this.url) ? 'utf8' : 'base64';
        fs_1.default.readFile(this.url, { encoding }, (err, data) => {
            if (err)
                throw err;
            this.response = data;
            this.responseText = data;
            const event = { target: { status: this.status } };
            this.onload(this, event);
        });
    }
    onload(xhr, event) { }
    onerror(err) { }
    onprogress() { }
}
exports.default = FakeXMLHttpRequest;
